File: workspace/scudoku/src/main/scala/scudoku/data/Grid.scala
package scudoku.data

/**
 * Java2Scala port from judoku.data.Grid
 */

import java.util.Observer
import java.util.Observable
import UnknownOrOneToNine._
import scudoku.Debug

class Grid extends Observer {
  private val NINE = 9
  
  // the lines of the grid
  val lines: Array[Line]     = new Array[Line](NINE)
  // the columns of the grid
  val columns: Array[Column] = new Array[Column](NINE)
  // the blocks of the grid
  val blocks: Array[Block]   = new Array[Block](NINE)
  // all cells of the grid
  val cells: Array[Cell]     = new Array[Cell](NINE*NINE)
  
  // number of cells which are fix already
  private var _numOfCellsFixed: Int = 0
  
  // helper cell id array
  private val ids = Array(
        Array( 0, 1, 2,  9,10,11, 18,19,20),
        Array( 3, 4, 5, 12,13,14, 21,22,23),
        Array( 6, 7, 8, 15,16,17, 24,25,26),
        Array(27,28,29, 36,37,38, 45,46,47),
        Array(30,31,32, 39,40,41, 48,49,50),
        Array(33,34,35, 42,43,44, 51,52,53),
        Array(54,55,56, 63,64,65, 72,73,74),
        Array(57,58,59, 66,67,68, 75,76,77),
        Array(60,61,62, 69,70,71, 78,79,80)
  )
  
  private val id2LineAndColumnMap = scala.collection.mutable.Map[Int, (Line, Column)]()
  
  // Constructor
  {
      // construct the lines, columns and blocks
      for (l <- 0 to lines.length-1)   lines(l)   = new Line()
      for (c <- 0 to columns.length-1) columns(c) = new Column()
      for (b <- 0 to blocks.length-1)  blocks(b)  = new Block()
      
      // create cells and link them to the correct block, line and col and also keep them in the map
      for (idsi <- 0 to ids.length-1) {
        for (idsj <- 0 to ids(idsi).length-1) {
          val id: Int = ids(idsi)(idsj)
          
          val myline: Line = lines(idsi)
          val mycolumn: Column = columns(idsj)
          
          id2LineAndColumnMap += (id -> (myline,mycolumn))
        }
      }
      
      for (id <- 0 to cells.length-1) {
        val block: Block = blocks(id/NINE)
        val line: Line = id2LineAndColumnMap(id)._1
        val column: Column = id2LineAndColumnMap(id)._2

        val cell: Cell = new Cell(this, block, line, column, id)
        cells(id) = cell
        
        block.addCell(cell)
        line.addCell(cell)
        column.addCell(cell)
        
        // set grid as observer to cell's changes
        cell.addObserver(this)
      }
  }
  
  // parse a given "Stringified" grid
  def parseStringGrid(sGrid: String) = {
    if (sGrid.length != NINE*NINE) throw new RuntimeException("grid size is not 9x9");
    var stringIndex: Int = 0
    for (idsi <- 0 to ids.length-1) {
      for (idsj <- 0 to ids(idsi).length-1) {
        val id: Int = ids(idsi)(idsj)
        val block: Block = blocks(id/NINE)
        
        val cell: Cell = block.getCell(id)

        val value = UnknownOrOneToNine.fromChar(sGrid.charAt(stringIndex))
        if (value != UnknownOrOneToNine.UNKNOWN) cell.value_=(value)

        stringIndex = stringIndex+1
      }
    }
  }

  // grid is observer to *all* cells so we get the number of fixed cells for free
  override def update(o: Observable, arg: Any) = {
    o match {
      case c: Cell if c.value != UNKNOWN => {
        _numOfCellsFixed = _numOfCellsFixed+1
      }
    }
  }
  
  // are we done?
  def isSolved = _numOfCellsFixed == NINE*NINE
  
  // get the number of cells fixed
  def numOfCellsFixed = _numOfCellsFixed

  def getAllCellSets(): List[CellSet] = {
    blocks.toList ::: lines.toList ::: columns.toList 
  }
  
  // ---------------------------------------------------------------------------

  override def toString: String = {
    val buf = new StringBuffer()
    
    var i: Int = 0
    for (l <- lines) {
      if (Debug.GRID_DEBUG) {
          buf.append("Line " + (if (i<10) " " else "") + i + ": ");
      }
      
      buf.append(l.toString());
      buf.append("\n");
      i = i+1
    }

    if (Debug.GRID_DEBUG) {
        i=0;
        for (b <- blocks) {
            buf.append("Block " + (if (i<10) " " else "") + i + ":\n");
            buf.append(b.toString());
            buf.append("\n");
            i = i+1
        }

        i=0;
        for (c <- columns) {
            buf.append("Column " + (if (i<10) " " else "") + i + ":\n");
            buf.append(c.toString());
            buf.append("\n");
            i = i+1
        }
    }
    
    buf.toString()
  }
}

File: workspace/judoku/src/main/java/judoku/data/Grid.java
package judoku.data;

import java.util.HashMap;
import java.util.Observable;
import java.util.Observer;
import java.util.Set;

import judoku.Debug;

public class Grid implements Observer {
	public static final int NINE = 9;
	
	// the lines of the grid
	private Line[]   lines   = new Line[NINE];
	// the columns of the grid
    private Column[] columns = new Column[NINE];
    // the blocks of the grid
    private Block[]  blocks  = new Block[NINE];
    // all cells in the grid
    private Cell[]   cells   = new Cell[NINE*NINE];

    // number of cells which are fix already
    private int numOfCellsFixed = 0;

    // helper map and cell id array
    private HashMap<Integer, Cell> cellsByID = new HashMap<Integer, Cell>();
    private int[][] ids =	// id's for the cell position, needed to construct the grid
	    { { 0, 1, 2,  9,10,11, 18,19,20},
	      { 3, 4, 5, 12,13,14, 21,22,23},
	      { 6, 7, 8, 15,16,17, 24,25,26},
	      {27,28,29, 36,37,38, 45,46,47},
	      {30,31,32, 39,40,41, 48,49,50},
	      {33,34,35, 42,43,44, 51,52,53},
	      {54,55,56, 63,64,65, 72,73,74},
	      {57,58,59, 66,67,68, 75,76,77},
	      {60,61,62, 69,70,71, 78,79,80} };

    public Grid() {
		// construct the lines, columns and blocks
        for (int l=0; l<lines.length;   l++)   lines[l] = new   Line(l);
        for (int c=0; c<columns.length; c++) columns[c] = new Column(c);
        for (int b=0; b<blocks.length;  b++)  blocks[b] = new  Block(b);

        // create cells and link them to the correct block, line and col and also keep them in the map
        for (int i=0; i<cells.length; i++) {
            Cell cell = new Cell(this, i);
            int id = cell.id;
            cells[i] = cell;

            // add cell to its block
            blocks[id/NINE].addCell(cell);  
            cell.setBlock(blocks[id/NINE]);
            
            // add cell to my map
            if (cellsByID.get(id) != null) throw new RuntimeException("cell duplicate found for id: " + id); // paranoia
            cellsByID.put(id, cells[i]);

            // set grid as observer to cell's changes
            cell.addObserver(this);
        }

        // finally set cell's line and column
        for (int idsi=0; idsi<ids.length; idsi++) {
            for (int idsj=0; idsj<ids[idsi].length; idsj++) {
                int id = ids[idsi][idsj];
                Cell cell = cellsByID.get(id);

                lines[idsi].addCell(cell);  	
                cell.setLine(lines[idsi]);
                columns [idsj].addCell(cell);  	
                cell.setColumn(columns [idsj]);
            }
        }
    }

    // parse a given "Stringified" grid
    public void parseStringGrid(String sGrid) {
        if (sGrid.length() != NINE*NINE) throw new RuntimeException("grid size is not 9x9");

        // iterate over the stringified grid and set all values
        int counter=0;
        for (int idsi=0; idsi<ids.length; idsi++) {
            for (int idsj=0; idsj<ids[idsi].length; idsj++) {
                int id = ids[idsi][idsj];
                Cell cell = cellsByID.get(id);
                cell.parseCharCell(sGrid.charAt(counter++));
            }
        }
    }

	// -----------------------------------------------------------------------------------------------

    public Block[] getBlocks() {
		return blocks;
	}
	public Column[] getColumns() {
		return columns;
	}
	public Line[] getLines() {
		return lines;
	}
    public Cell[] getCells() {
        return cells;
    }

	// -----------------------------------------------------------------------------------------------

    // grid is observer to *all* cells so we get the number of fixed cells for free
    @Override
    public void update (Observable obs, Object obj) {
        if (obs== null) return;
        Cell oAsCell = (Cell) obs;
        if (oAsCell.getValue() != null) {
            numOfCellsFixed++;
        }
    }

	// -----------------------------------------------------------------------------------------------

    // are we done?
    public boolean isSolved() {
        return (numOfCellsFixed == NINE*NINE);
    }

    // get the number of cells fixed
    public int numOfCellsFixed() {
        return this.numOfCellsFixed;
    }

	// -----------------------------------------------------------------------------------------------

    public String toString() {
        StringBuffer buf = new StringBuffer();

        int i=0;

        for (Line l: lines) {
            if (Debug.GRID_DEBUG) {
                buf.append("Line " + ((i<10)? " ": "")+ (++i)+ ": ");
            }

            buf.append(l.toString());
            buf.append("\n");
        }

        if (Debug.GRID_DEBUG) {
            i=0;
            for (Block b: blocks) {
                buf.append("Block " + ((i<10)? " ": "")+ (++i) + ":\n");
                buf.append(b.toString());
                buf.append("\n");
            }

            i=0;
            for (Column c: columns) {
                buf.append("Column " + ((i<10)? " ": "")+ (++i) + ":\n");
                buf.append(c.toString());
                buf.append("\n");
            }
        }
        return buf.toString();
   }

   public Object clone() {
	   Grid clone = new Grid();

	   // iterate over all cells of the original grid
	   for (Cell orgCell: this.getCells()) {
		   // find the corresponding cell in the clone grid
		   Cell cloneCell = clone.cellsByID.get(orgCell.id);

		   // if the original cell is fixed
		   if (orgCell.isFix()) {
			   // ... simply set the value
			   cloneCell.setValue(orgCell.getValue());
		   }
		   else {
			   // ... else iterate over all possible values and delete in the clone
			   // whatever cannot be found in the original cell
			   Set<OneToNine> orgPotentialValues = orgCell.getPotentialValues();
			   for (OneToNine o2n: OneToNine.values()) {
				   if (!orgPotentialValues.contains(o2n)) {
					   cloneCell.deletePotentialValue(o2n);
				   }
			   }
		   }
	   }

	   return clone;
   }
}